# Cleaning the Current_inventory Excel Sheet (CSV)...Again.

Why? because this time I have a better understanding of the assingment (AT LAST!).

In [14]:
from pathlib import Path
import pandas as pd

In [15]:
# Dynamically get the project root (where your notebook is)
notebook_dir = Path().resolve()

# Then navigate to the CSV
file_path = notebook_dir.parent / "raws" / "current_inventory.csv"

# Confirm the path exists before loading
print("Looking for:", file_path)
assert file_path.exists(), "CSV file not found at that location!"


# Read the CSV
df = pd.read_csv(file_path)
df.head(10)

Looking for: C:\Users\dell\Documents\desktop\real projects\CTSE_ASSINGMENT\raws\current_inventory.csv


,1009,1009.1,6655,Unnamed: 3,5745,0,-9,-1,-4,0.1,-9.1,-2,0.2,0.3,-4.1,0.4,0.5,0.6,0.7
0,Product Name,msku,Opening Stock,Buffer Stock,TLCQ,BLR7,BLR8,BOM5,BOM7,CCU1,CCX1,DEL4,DEL5,DEX3,PNQ2,PNQ3,SDED,SDEE,XHJ9
1,Chimmy Pillow,CSTE_0001_ST_Bts_Pillow_Chimmy,192,NaN,192,0,0,0,-1,0,0,0,0,0,0,0,0,0,0
2,Tata Pillow,CSTE_0002_ST_Bts_Pillow_Tata,0,NaN,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
3,Koya Pillow,CSTE_0003_ST_Bts_Pillow_Koya,0,NaN,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
4,Mang Pillow,CSTE_0004_ST_Bts_Pillow_Mang,21,NaN,21,0,0,0,0,0,0,0,0,0,0,0,0,0,0
5,Shooky Pillow,CSTE_0005_ST_Bts_Pillow_Shooky,0,NaN,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
6,Van Pillow,CSTE_0006_ST_Bts_Pillow_Van,49,NaN,49,0,0,0,0,0,0,0,0,0,0,0,0,0,0
7,Cooky Pillow,CSTE_0007_ST_Bts_Pillow_Cooky,0,NaN,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
8,Rj Pillow,CSTE_0008_ST_Bts_Pillow_Rj,0,NaN,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
9,Chimmy Long Pillow,CSTE_0009_ST_Bts_LongPillow_Chimmy,0,NaN,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0


In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1012 entries, 0 to 1011
Data columns (total 19 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   1009        1012 non-null   object
 1   1009.1      1012 non-null   object
 2   6655        1012 non-null   object
 3   Unnamed: 3  8 non-null      object
 4   5745        1012 non-null   object
 5   0           1012 non-null   object
 6   -9          1012 non-null   object
 7   -1          1012 non-null   object
 8   -4          1012 non-null   object
 9   0.1         1012 non-null   object
 10  -9.1        1012 non-null   object
 11  -2          1012 non-null   object
 12  0.2         1012 non-null   object
 13  0.3         1012 non-null   object
 14  -4.1        1012 non-null   object
 15  0.4         1012 non-null   object
 16  0.5         1012 non-null   object
 17  0.6         1012 non-null   object
 18  0.7         1012 non-null   object
dtypes: object(19)
memory usage: 150.3+ KB


## Step 1:Examine the data types and get basic info about the dataframe
### Questions to be answered:
1. What data types pandas has assigned to each column.
2. Whether there are any null values
3. The exact structure of the first few rows

In [17]:
# Check the data types and basic info
print("Data types:")
print(df.dtypes)
print("\n" + "="*50 + "\n")

print("DataFrame info:")
print(df.info())
print("\n" + "="*50 + "\n")

print("First few rows to see the structure:")
print(df.head(3))

Data types:
1009          object
1009.1        object
6655          object
Unnamed: 3    object
5745          object
0             object
-9            object
-1            object
-4            object
0.1           object
-9.1          object
-2            object
0.2           object
0.3           object
-4.1          object
0.4           object
0.5           object
0.6           object
0.7           object
dtype: object


DataFrame info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1012 entries, 0 to 1011
Data columns (total 19 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   1009        1012 non-null   object
 1   1009.1      1012 non-null   object
 2   6655        1012 non-null   object
 3   Unnamed: 3  8 non-null      object
 4   5745        1012 non-null   object
 5   0           1012 non-null   object
 6   -9          1012 non-null   object
 7   -1          1012 non-null   object
 8   -4          1012 non-null   object
 9   0.1  

### The main problems
1. **Column names are wrong** - they're showing the summary numbers instead of proper names
2. **The actual column names are in the first row** - (Product Name, msku, Opening Stock, etc.)
3. **All data is stored as 'object' type** - instead of proper numeric types.
4. **The first row contains headers, not data**



## Step 2: Fixing the Column Names and removing the header row


In [18]:
# First, let's see what the actual column names should be (they're in row 0)
print("Current first row (which contains our real column names):")
print(df.iloc[0].tolist())
print("\n" + "="*50 + "\n")

# Set the first row as column names and remove it from the data
new_column_names = df.iloc[0].tolist()
print("New column names will be:")
print(new_column_names)
print("\n" + "="*50 + "\n")

# Create a new dataframe with proper column names
df_cleaned = df.iloc[1:].copy()  # Take all rows except the first one
df_cleaned.columns = new_column_names  # Set the proper column names

# Reset the index so it starts from 0 again
df_cleaned = df_cleaned.reset_index(drop=True)

print("After fixing column names:")
print("Shape:", df_cleaned.shape)
print("\nFirst few rows:")
print(df_cleaned.head(3))

Current first row (which contains our real column names):
['Product Name', 'msku', 'Opening Stock', 'Buffer Stock', 'TLCQ', 'BLR7', 'BLR8', 'BOM5', 'BOM7', 'CCU1', 'CCX1', 'DEL4', 'DEL5', 'DEX3', 'PNQ2', 'PNQ3', 'SDED', 'SDEE', 'XHJ9']


New column names will be:
['Product Name', 'msku', 'Opening Stock', 'Buffer Stock', 'TLCQ', 'BLR7', 'BLR8', 'BOM5', 'BOM7', 'CCU1', 'CCX1', 'DEL4', 'DEL5', 'DEX3', 'PNQ2', 'PNQ3', 'SDED', 'SDEE', 'XHJ9']


After fixing column names:
Shape: (1011, 19)

First few rows:
     Product Name                            msku Opening Stock Buffer Stock  \
0  Chimmy Pillow   CSTE_0001_ST_Bts_Pillow_Chimmy           192          NaN   
1    Tata Pillow     CSTE_0002_ST_Bts_Pillow_Tata             0          NaN   
2    Koya Pillow     CSTE_0003_ST_Bts_Pillow_Koya             0          NaN   

  TLCQ BLR7 BLR8 BOM5 BOM7 CCU1 CCX1 DEL4 DEL5 DEX3 PNQ2 PNQ3 SDED SDEE XHJ9  
0  192    0    0    0   -1    0    0    0    0    0    0    0    0    0    0  
1    0    0    

## Why are there still 1011 rows?

I know one row is the head row but did we still include the summary row in the data?

If there are supposed to be 1009 products but we have 1011 rows after removing the header, that means there are likely 2 extra rows that aren't actual product data.

Let's investigate. 


### Step 2.5:  Check for summary or non-product rows

In [19]:
# Check for any rows where Product Name might be a number (summary rows)
print("Checking for rows where Product Name contains only numbers:")
print("\n" + "="*50 + "\n")

# Check if any Product Name values look like numbers
numeric_product_names = []
for idx, product_name in enumerate(df_cleaned['Product Name']):
    # Convert to string and strip whitespace, then check if it's numeric
    str_name = str(product_name).strip()
    if str_name.isdigit() or str_name.replace('.', '').isdigit():
        numeric_product_names.append((idx, product_name))

print("Rows with numeric Product Names:")
if numeric_product_names:
    for idx, name in numeric_product_names:
        print(f"Row {idx}: Product Name = '{name}'")
        print(df_cleaned.iloc[idx])
        print("-" * 30)
else:
    print("No numeric Product Names found")

print("\n" + "="*50 + "\n")

# Also check the very end for any potential summary rows
print("Let's examine the last few rows more carefully:")
for i in range(max(0, len(df_cleaned)-5), len(df_cleaned)):
    row = df_cleaned.iloc[i]
    print(f"Row {i}: '{row['Product Name']}' | MSKU: '{row['msku']}' | Opening Stock: '{row['Opening Stock']}'")

Checking for rows where Product Name contains only numbers:


Rows with numeric Product Names:
No numeric Product Names found


Let's examine the last few rows more carefully:
Row 1006: 'USB Hub Adaptor' | MSKU: 'XEXY_0004_Accs_Electronics_USB_Hub' | Opening Stock: '0'
Row 1007: 'Tripod K31' | MSKU: 'XEXY_0005_Accs_Electronics_Tripod_K31' | Opening Stock: '1'
Row 1008: 'Lightning OTG Adapter' | MSKU: 'XEXY_0006_Accs_Electronics_OTG_Adapter' | Opening Stock: '47'
Row 1009: 'Breathing Otter Grey' | MSKU: 'CSTE_0549_ST_Animal_Breathing_Otter_Grey' | Opening Stock: '14'
Row 1010: 'Kangaroo plush toy' | MSKU: 'CSTE_O0447_OT_Kangaroo_Plush' | Opening Stock: '1'


### Findings

1. **No numeric Product Names found** - So no summary rows to remove
2. **All rows at the end look legitimate** - proper product names and MSKUs 
3. **The "1009" in the original summary was incorrect** - CSTE actually has **1011 products** in it's inventory not 1009

**1011 rows are valid product data, and we don't need to remove any rows. The discrepancy was in the original summary information.**

## Step 3: Fix the data types 

All columns are stored as 'object' (strings), we need to convert the numeric columns to proper integer types

In [20]:
# First, let's verify all MSKUs are unique (they should be)
print("Verifying MSKU uniqueness:")
print(f"Total rows: {len(df_cleaned)}")
print(f"Unique MSKUs: {df_cleaned['msku'].nunique()}")
print("All MSKUs are unique:", len(df_cleaned) == df_cleaned['msku'].nunique())
print("\n" + "="*50 + "\n")

# Replace NaN in Buffer Stock with 0 as mentioned
df_cleaned['Buffer Stock'] = df_cleaned['Buffer Stock'].fillna(0)

# Convert numeric columns to integers
numeric_columns = ['Opening Stock', 'Buffer Stock'] + [col for col in df_cleaned.columns if col not in ['Product Name', 'msku']]

print("Converting these columns to integers:")
print(numeric_columns)
print("\n" + "="*30 + "\n")

# Convert to integers
for col in numeric_columns:
    df_cleaned[col] = pd.to_numeric(df_cleaned[col], errors='coerce').astype(int)

print("After conversion - Data types:")
print(df_cleaned.dtypes)
print("\nFirst few rows with proper data types:")
print(df_cleaned.head(3))

Verifying MSKU uniqueness:
Total rows: 1011
Unique MSKUs: 1011
All MSKUs are unique: True


Converting these columns to integers:
['Opening Stock', 'Buffer Stock', 'Opening Stock', 'Buffer Stock', 'TLCQ', 'BLR7', 'BLR8', 'BOM5', 'BOM7', 'CCU1', 'CCX1', 'DEL4', 'DEL5', 'DEX3', 'PNQ2', 'PNQ3', 'SDED', 'SDEE', 'XHJ9']


After conversion - Data types:
Product Name     object
msku             object
Opening Stock     int64
Buffer Stock      int64
TLCQ              int64
BLR7              int64
BLR8              int64
BOM5              int64
BOM7              int64
CCU1              int64
CCX1              int64
DEL4              int64
DEL5              int64
DEX3              int64
PNQ2              int64
PNQ3              int64
SDED              int64
SDEE              int64
XHJ9              int64
dtype: object

First few rows with proper data types:
     Product Name                            msku  Opening Stock  \
0  Chimmy Pillow   CSTE_0001_ST_Bts_Pillow_Chimmy            192   
1   

# The Data Cleaning is now **Complete**

## Data Cleaning Summary:
1. **Fixed column names** - Converted from meaningless numbers to proper names
2. **Verified data integrity** - All 1011 products have unique MSKUs
3. **Corrected data types** - All numeric columns now properly stored as integers
4. **Handled missing values** - Replaced NaN in Buffer Stock with 0
5. **Confirmed no summary rows** - All rows are legitimate product data

## Step 4: Data Overview

In [21]:
# Let's get a final overview of our clean data
print("=== FINAL CLEANED DATASET OVERVIEW ===")
print(f"Shape: {df_cleaned.shape}")
print(f"Total Products: {len(df_cleaned)}")
print("\nColumn Summary:")
print("Text columns:", ['Product Name', 'msku'])
warehouse_cols = [col for col in df_cleaned.columns if col not in ['Product Name', 'msku', 'Opening Stock', 'Buffer Stock']]
print("Warehouse columns:", warehouse_cols)

print("\n=== STOCK SUMMARY ===")
print("Opening Stock:")
print(f"  Total Opening Stock: {df_cleaned['Opening Stock'].sum():,}")
print(f"  Products with stock > 0: {(df_cleaned['Opening Stock'] > 0).sum()}")
print(f"  Products with zero stock: {(df_cleaned['Opening Stock'] == 0).sum()}")

print("\nBuffer Stock:")
print(f"  Total Buffer Stock: {df_cleaned['Buffer Stock'].sum():,}")

print("\nWarehouse Stock (TLCQ to XHJ9):")
warehouse_total = df_cleaned[warehouse_cols].sum().sum()
print(f"  Total across all warehouses: {warehouse_total:,}")
print(f"  Individual warehouse totals:")
for col in warehouse_cols:
    warehouse_sum = df_cleaned[col].sum()
    print(f"    {col}: {warehouse_sum:,}")

# Check for negative values (backorders)
print("\n=== NEGATIVE STOCK CHECK ===")
negative_opening = (df_cleaned['Opening Stock'] < 0).sum()
negative_warehouses = sum((df_cleaned[col] < 0).sum() for col in warehouse_cols)
print(f"Products with negative Opening Stock: {negative_opening}")
print(f"Warehouse locations with negative stock: {negative_warehouses}")

print("\nFirst 3 rows of clean data:")
print(df_cleaned.head(3))

# Save the cleaned DataFrame to a new CSV file
output_dir = notebook_dir.parent / "clean_data"
output_dir.mkdir(parents=True, exist_ok=True)
output_file = output_dir / "cleaned_inventory.csv"

df_cleaned.to_csv(output_file, index=False)

=== FINAL CLEANED DATASET OVERVIEW ===
Shape: (1011, 19)
Total Products: 1011

Column Summary:
Text columns: ['Product Name', 'msku']
Warehouse columns: ['TLCQ', 'BLR7', 'BLR8', 'BOM5', 'BOM7', 'CCU1', 'CCX1', 'DEL4', 'DEL5', 'DEX3', 'PNQ2', 'PNQ3', 'SDED', 'SDEE', 'XHJ9']

=== STOCK SUMMARY ===
Opening Stock:
  Total Opening Stock: 6,670
  Products with stock > 0: 348
  Products with zero stock: 663

Buffer Stock:
  Total Buffer Stock: 350

Warehouse Stock (TLCQ to XHJ9):
  Total across all warehouses: 5,726
  Individual warehouse totals:
    TLCQ: 5,755
    BLR7: 0
    BLR8: -9
    BOM5: -1
    BOM7: -4
    CCU1: 0
    CCX1: -9
    DEL4: -2
    DEL5: 0
    DEX3: 0
    PNQ2: -4
    PNQ3: 0
    SDED: 0
    SDEE: 0
    XHJ9: 0

=== NEGATIVE STOCK CHECK ===
Products with negative Opening Stock: 0
Warehouse locations with negative stock: 45

First 3 rows of clean data:
     Product Name                            msku  Opening Stock  \
0  Chimmy Pillow   CSTE_0001_ST_Bts_Pillow_Chimmy    

## **Key Insights from the Clean Data:**

✅ **Opening Stock Validation**: Correctly shows 0 products with negative opening stock (as it should be)

✅ **Inventory Distribution**:
- **6,670 total units** in opening stock
- Only **348 out of 1011 products** (34%) currently have stock
- **663 products** (66%) are out of stock

✅ **Warehouse Analysis**:
- **TLCQ is the main warehouse** with 5,755 units (85% of all warehouse stock)
- **45 backorder situations** across warehouses (negative stock = oversold/promised but not available)
- Most other warehouses are empty or have minimal stock

## **About Buffer Stock- What is Buffer Stock?**
Buffer Stock is typically "safety stock" or "reserve stock" - extra inventory kept on hand to prevent stockouts due to demand fluctuations or supply delays. CSTE's total of 350 units suggests a conservative buffer strategy.